### PR TITLE
feat: ignore file name recursively

### LINKED ISSUE #158: Exclude special filetypes from Include via Ignore section in spec.yaml
### TL;DR

An idea was proposed to add a separate .specignore file for excluding special system files (like .DS_Store) or other private files that should not be included in template renders. Having a separate file for this would require template readers to have to go back and forth between spec.yaml and .specignore to understand how the render should behave. Instead, this could be included as a separate optional "ignore" section in the spec.yaml. Adding either a .specignore or an ignore section in spec.yaml would also support nice-to-haves such as a future IDE plugin that could detect and gray out such files.

It was decided to hold off on this feature until we have a customer who needs this functionality.

An example of this might look like:
```
apiVersion: 'cli.abcxyz.dev/v1alpha1'
kind: 'Template'
desc: 'Example template'

ignore: ['.DS_Store, '.bin', '.ssh']

inputs:
steps:
...
```

[proposal for this feature in Include Enhancements doc](https://docs.google.com/document/d/12iN2g1Idz6fW5ZqaokTmSFZf5kHXaWeewULWy2q5GfE/edit#heading=h.2wia9q1hlzad)

Idea from @sethvargo.

### Detailed design

_No response_

### Alternatives considered

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by sqin2019**
fix #158

### CODE REVIEW

File: templates/commands/render/action_include_test.go [hunk 1]
```diff
@@ -443,30 +443,33 @@ func TestActionInclude(t *testing.T) {
...
+						Paths: modelStrings([]string{"."}),
 					},
 					{
 						Paths: modelStrings([]string{"."}),
 						From:  model.String{Val: "destination"},
 					},
 				},
 			},
-			ignorePatterns: modelStrings([]string{"folder1/folder2", "file2.txt"}),
+			ignorePatterns: modelStrings([]string{"folder1/folder2", "file1.txt", "*.cfg"}),
```
Comment (drevell) [inline templates/commands/render/action_include_test.go:454]: I don't think we have a test for ignoring a single named file without globs that's in the root directory (not a subdirectory). Could you add one?

**Review by drevell — CHANGES_REQUESTED** [review]

File: templates/commands/render/action_include_test.go (hunk 1 above)
Comment (sqin2019) [reply templates/commands/render/action_include_test.go:454]: Added line 457, let me know if you prefer separating them into separate test cases.

Comment (drevell) [reply templates/commands/render/action_include_test.go:454]: I think we might need still more tests. From your TODO:

```
// TODO(#158): add support for ignoring file name. For example, "/foo" ignores
// a single file named foo in the root, but "foo" ignores all files named foo
// anywhere.
```

Things we probably want tests for:
 - `/subdir/file` (file in subdir with leading slash)
 - `file` ignores file in subdir and in root
 - `/file` ignores file in root but not in subdir
 - `*.cfg` ignores files in root and in subdir
 - `/*.cfg` ignores files in root but not in subdir
 - `/*/*.cfg` ignores files in subdir but not in root and not in some other subdir like `subdir1/subdir2/foo.cfg`
 - `subdir1/subdir2` ignores `othersubdir/subdir1/subdir2`
 - `/subdir1/subdir2` ignores `/subdir1/subdir2` but not `othersubdir/subdir1/subdir2`
 
 So.. basically the combinations of (1) with/without leading slash (2) file vs directory (3) with/without globs (4) file/dir being ignored is in root vs in subdirectory

Comment (sqin2019) [reply templates/commands/render/action_include_test.go:454]: (1) and (4) are the same? One thing is missing in the TODO is that currently we are matching files/dirs based on their path relative to the template dir or destination dir if it is from destination, I am thinking not supporting leading slash is fine?

File: templates/commands/render/action_include.go [hunk 2]
```diff
@@ -161,7 +161,15 @@ func checkIgnore(patterns []model.String, path string) (bool, error) {
 		patterns = defaultIgnorePatterns
```
Comment (sqin2019) [inline templates/commands/render/action_include.go:161]: Note: remove TODO

File: templates/commands/render/action_include_test.go (hunk 1 above)
Comment (sqin2019) [reply templates/commands/render/action_include_test.go:454]: There is one edge case we cannot handle if user want to only ignore a file under destination dir without accidentally remove files with the same name in subdirs if we are using all relative path patterns. Maybe we do need to support leading slash for this, but I am still thinking if there are other workarounds...

Comment (sqin2019) [reply templates/commands/render/action_include_test.go]: Updated to support leading slash.

File: templates/commands/render/action_include.go [hunk 3]
```diff
@@ -151,17 +151,25 @@ func includePath(ctx context.Context, inc *spec.IncludePath, sp *stepParams) err
...
 	for _, p := range patterns {
-		matched, err := filepath.Match(filepath.FromSlash(p.Val), path)
+		var matched bool
+		var err error
+		if filepath.Base(p.Val) == p.Val {
+			// Match file name if the pattern value is file name instead of path.
+			matched, err = filepath.Match(p.Val, filepath.Base(path))
+		} else if p.Val[0] == '/' {
+			// Match absolute path if the pattern start with a leading slash.
+			matched, err = filepath.Match(filepath.Join(fromDir, filepath.FromSlash(p.Val)), filepath.Join(fromDir, path))
```
Comment (drevell) [inline templates/commands/render/action_include.go:168]: Check my math on this, but I think this line can be simplified to:

```
matched, err = filepath.Match(filepath.FromSlash(p.Val[1:]), path)
```

**Review by drevell — CHANGES_REQUESTED** [review]
Only some nits

File: templates/commands/render/action_include.go (hunk 3 above)
Comment (sqin2019) [reply templates/commands/render/action_include.go:168]: You are right, I thought it is going to match the subdirs too but I was wrong. updating, thanks!

**Review by drevell — APPROVED** [review]

**Pull request merged**